boa/src/builtins/date/mod.rs

```diff
@@ -0,0 +1,1421 @@
+            fn get_arg(i: usize, args: &[Value], ctx: &mut Interpreter) -> Option<f64> {
+                args
+                .get(i)
+                .and_then(|value| {
+                    ctx.to_numeric_number(value).map_or_else(
+                        |_| None,
+                        |value| {
+                            if value == 0f64 || value.is_normal() {
+                                Some(value)
+                            } else {
+                                None
+                            }
+                        },
+                    )
+                })
+            }
+
+            let mut $var = [None; $count];
+            for i in 0..$count {
+                $var[i] = get_arg(i, args, ctx);
+            }
+
+            let inner = Date::this_time_value(this, ctx)?.$tz();
+            let new_value = inner.and_then(|$date_time| $mutate);
+            let new_value = new_value.map(|date_time| date_time.naive_utc());
+            this.set_data(ObjectData::Date(Date(new_value)));
+
+            Ok(Value::number(
+                Self::this_time_value(this, ctx)?.timestamp(),
+            ))
+        }
+    };
+    ($(#[$outer:meta])* fn $name:ident ($var:ident[$count:literal]) $mutate:expr) => {
+        $(#[$outer])*
+        fn $name(this: &Value, args: &[Value], ctx: &mut Interpreter) -> ResultValue {
+            // If the first arg is not present or NaN, the Date becomes NaN itself.
+            fn get_arg(i: usize, args: &[Value], ctx: &mut Interpreter) -> Option<f64> {
+                args
+                .get(i)
+                .and_then(|value| {
+                    ctx.to_numeric_number(value).map_or_else(
+                        |_| None,
+                        |value| {
+                            if value == 0f64 || value.is_normal() {
+                                Some(value)
+                            } else {
+                                None
+                            }
+                        },
+                    )
+                })
+            }
+
+            let mut $var = [None; $count];
+            for i in 0..$count {
+                $var[i] = get_arg(i, args, ctx);
+            }
+
+            let new_value = $mutate;
+            let new_value = new_value.map(|date_time| date_time.naive_utc());
+            this.set_data(ObjectData::Date(Date(new_value)));
+
+            Ok(Value::number(
+                Self::this_time_value(this, ctx)?.timestamp(),
+            ))
+        }
+    };
+}
+
+#[derive(Debug, Finalize, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Date(Option<NaiveDateTime>);
+
+impl Display for Date {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.to_local() {
+            Some(v) => write!(f, "{}", v.format("%a %b %d %Y %H:%M:%S GMT%:z")),
+            _ => write!(f, "Invalid Date"),
+        }
+    }
+}
+
+unsafe impl Trace for Date {
```

Comment:
I would add a comment here, explaining that this is only safe if `chrono` never implements `Trace` for `NaiveDateTime`.